cpp/src/arrow/compute/kernels/util-internal.cc

```diff
@@ -165,46 +171,96 @@ Datum WrapDatumsLike(const Datum& value, const std::vector<Datum>& datums) {
 }
 
 PrimitiveAllocatingUnaryKernel::PrimitiveAllocatingUnaryKernel(
-    std::unique_ptr<UnaryKernel> delegate)
-    : delegate_(std::move(delegate)) {}
+    UnaryKernel* delegate, const std::shared_ptr<DataType>& out_type)
+    : delegate_(delegate), out_type_(out_type) {}
+
+PrimitiveAllocatingUnaryKernel::PrimitiveAllocatingUnaryKernel(
+    std::unique_ptr<UnaryKernel> delegate, const std::shared_ptr<DataType>& out_type)
+    : PrimitiveAllocatingUnaryKernel(delegate.get(), out_type) {
+  owned_delegate_ = std::move(delegate);
+}
 
 inline void ZeroLastByte(Buffer* buffer) {
   *(buffer->mutable_data() + (buffer->size() - 1)) = 0;
 }
 
-Status PrimitiveAllocatingUnaryKernel::Call(FunctionContext* ctx, const Datum& input,
-                                            Datum* out) {
-  std::vector<std::shared_ptr<Buffer>> data_buffers;
-  const ArrayData& in_data = *input.array();
-  MemoryPool* pool = ctx->memory_pool();
+Status PropagateNulls(FunctionContext* ctx, const ArrayData& input, ArrayData* output) {
+  const int64_t length = input.length;
 
-  // Handle the validity buffer.
-  if (in_data.offset == 0 || in_data.null_count <= 0) {
-    // Validity bitmap will be zero copied (or allocated when buffer is known).
-    data_buffers.emplace_back();
-  } else {
+  if (output->buffers.size() == 0) {
+    // Ensure we can assign a buffer
+    output->buffers.resize(1);
+  }
+
+  // Handle validity bitmap
+  output->null_count = input.GetNullCount();
+  if (input.offset != 0 && output->null_count > 0) {
+    DCHECK(input.buffers[0]);
+    const Buffer& validity_bitmap = *input.buffers[0];
     std::shared_ptr<Buffer> buffer;
-    RETURN_NOT_OK(AllocateBitmap(pool, in_data.length, &buffer));
+    RETURN_NOT_OK(ctx->Allocate(BitUtil::BytesForBits(length), &buffer));
```

Comment:
just curious why the preference for this form vs the original?

Reply:
This way the memory pool doesn't have to leak. I don't have a strong preference

Reply:
Neither do I just curious about the change.